src/script/microbit-interfacing/state-updaters.ts

```diff
@@ -0,0 +1,188 @@
+      s.modelView = ModelView.TILE;
+      return s;
+    });
+  } else {
+    state.update(s => {
+      s.outputOrigin = HexOrigin.MAKECODE;
+      s.modelView = ModelView.TILE;
+      return s;
+    });
+  }
+};
+
+export const stateOnIdentifiedAsProprietary = (
+  requestState: DeviceRequestStates,
+): void => {
+  if (requestState === DeviceRequestStates.INPUT) {
+    state.update(s => {
+      s.inputOrigin = HexOrigin.PROPRIETARY;
+      s.modelView = ModelView.STACK;
+      return s;
+    });
+  } else {
+    state.update(s => {
+      s.outputOrigin = HexOrigin.PROPRIETARY;
+      s.modelView = ModelView.STACK;
+      return s;
+    });
+  }
+};
+
+export const stateOnReady = (requestState: DeviceRequestStates) => {
+  if (requestState === DeviceRequestStates.INPUT) {
+    // clearTimeout(this.reconnectTimeout);
+    state.update(s => {
+      s.isInputReady = true;
+      return s;
+    });
+    if (get(currentPath) === Paths.HOME) {
+      navigate(Paths.DATA);
+    }
+  } else {
+    // Reset any output pins currently active.
+    const pinResetArguments: { pin: MBSpecs.UsableIOPin; on: boolean }[] = [];
+    StaticConfiguration.supportedPins.forEach(pin => {
+      const argument = { pin: pin, on: false };
+      pinResetArguments.push(argument);
+    });
+    sendToOutput['sendToOutputPin'](pinResetArguments);
+    state.update(s => {
+      s.isOutputReady = true;
+      return s;
+    });
+    //clearTimeout(this.reconnectTimeout);
+  }
+};
+
+export const stateOnAssigned = (
+  requestState: DeviceRequestStates,
+  microbitVersion: MBSpecs.MBVersion,
+) => {
+  if (requestState === DeviceRequestStates.INPUT) {
+    state.update(s => {
+      s.isInputAssigned = true;
+      s.inputMicrobitVersion = microbitVersion;
+      return s;
+    });
+  } else {
+    state.update(s => {
+      s.isOutputAssigned = true;
+      s.inputMicrobitVersion = microbitVersion;
+      return s;
+    });
+  }
+  if (get(currentPath) === Paths.HOME) {
+    navigate(Paths.DATA);
+  }
+};
+
+export const stateOnDisconnected = (
+  requestState: DeviceRequestStates,
+  userDisconnect: boolean,
+): void => {
+  if (requestState === DeviceRequestStates.INPUT) {
+    state.update(s => {
+      s.isInputConnected = false;
+      // Hang on to the reference
+      // s.isInputAssigned = false;
+      s.isInputReady = false;
+      s.offerReconnect = !userDisconnect;
+      s.reconnectState = DeviceRequestStates.INPUT;
+      //s.isInputOutdated = false;
+      return s;
+    });
+  } else {
+    state.update(s => {
+      s.isOutputConnected = false;
+      s.offerReconnect = !userDisconnect;
+      // Hang on to the reference
+      // s.isOutputAssigned = false;
+      s.isOutputReady = false;
+      s.reconnectState = DeviceRequestStates.NONE;
+      //s.isOutputOutdated = false;
+      // TODO: Come back to this.
+      // if (!bothDisconnected) {
+      //   s.reconnectState = DeviceRequestStates.OUTPUT;
+      // }
```

Comment:
Flagging todo

Reply:
I think this is the correct behaviour [9b586b0](https://github.com/microbit-foundation/ml-trainer/pull/194/commits/9b586b0972af9a1fd43ce5c630b559cfbae4be7d).